src/main/java/com/mathworks/ci/actions/MatlabAction.java

```diff
@@ -0,0 +1,83 @@
+import com.mathworks.ci.BuildConsoleAnnotator;
+import com.mathworks.ci.MatlabBuilderConstants;
+import com.mathworks.ci.utilities.MatlabCommandRunner;
+import hudson.FilePath;
+import hudson.model.Run;
+import org.apache.commons.lang.RandomStringUtils;
+
+import java.io.File;
+import java.io.IOException;
+
+public class MatlabAction {
+    MatlabCommandRunner runner;
+    BuildConsoleAnnotator annotator;
+    String actionID;
+
+    public String getActionID(){
+        return (this.actionID == null) ? "" : this.actionID;
+    }
+
+    public MatlabAction(MatlabCommandRunner runner) {
+        this.runner = runner;
+    }
+
+    public MatlabAction(MatlabCommandRunner runner, BuildConsoleAnnotator annotator) {
+        this.runner = runner;
+        this.actionID = RandomStringUtils.randomAlphanumeric(8);
+        this.annotator = annotator;
+    }
+
+    public void copyBuildPluginsToTemp() throws IOException, InterruptedException {
+        // Copy plugins and override default plugins function
+        runner.copyFileToTempFolder(MatlabBuilderConstants.DEFAULT_PLUGIN, MatlabBuilderConstants.DEFAULT_PLUGIN);
+        runner.copyFileToTempFolder(MatlabBuilderConstants.BUILD_REPORT_PLUGIN, MatlabBuilderConstants.BUILD_REPORT_PLUGIN);
+        runner.copyFileToTempFolder(MatlabBuilderConstants.TASK_RUN_PROGRESS_PLUGIN, MatlabBuilderConstants.TASK_RUN_PROGRESS_PLUGIN);
+    }
+
+    public void setBuildEnvVars() throws IOException, InterruptedException {
+        // Set environment variable
+        runner.addEnvironmentVariable(
+                "MW_MATLAB_BUILDTOOL_DEFAULT_PLUGINS_FCN_OVERRIDE",
+                "ciplugins.jenkins.getDefaultPlugins");
+        runner.addEnvironmentVariable("MW_BUILD_PLUGIN_ACTION_ID",this.getActionID());
+        runner.addEnvironmentVariable(
+                "MW_MATLAB_TEMP_FOLDER",
+                runner.getTempFolder().toString());
+    }
+
+    public void teardownAction() {
+        try {
+            this.runner.removeTempFolder();
+        } catch (Exception e) {
+            System.err.println(e.toString());
+        }
+    }
+
+    public void moveJsonArtifactToBuildRoot(Run<?,?> build, String artifactBaseName) {
+        try {
+            FilePath file = new FilePath(runner.getTempFolder(), artifactBaseName + ".json");
+            if (file.exists()) {
+                FilePath rootLocation = new FilePath(
+                        new File(
+                                build.getRootDir().getAbsolutePath(),
+                                artifactBaseName + this.getActionID() + ".json")
+                );
+                file.copyTo(rootLocation);
+                file.delete();
+                build.addAction(new BuildArtifactAction(build, this.getActionID()));
+            }
+        } catch (Exception e) {
+            // Don't want to override more important error
+            // thrown in catch block
+            System.err.println(e.toString());
+        }
+    }

```

Comment:
Hmm, I'm a little bit nervous here about someone trying to call this method after using the constructor which does not provide an annotation. Could we check for the annotator being null and then throw an explicit error in that case rather than getting the null exception when it tries to concatenate `null` on line 66?

Reply:
Maybe I'm missing something but I'm unable to understand the concern here. Which part of line 66 are you expecting to equate to `null`? If I assume it's `runner.getTempFolder()`, all the Action class constructors require a runner and therefore `runner.getTempFolder` wouldn't give `null`.

P.S. For any other scenario as well if we need a `null` check, checking `null` for annotator would only solve for moving build artifact and **not** test results artifact (I mentioned in one of the previous comments about using this method for both build and test artifact in future)

Reply:
Sorry, you're totally right 66 is the wrong line number. I was thinking of the call to `this.getActionID()`, since that has the potential to return `null`. Might be worth handling just to make it harder to make mistakes in the codebase in the future like calling the wrong constructor.

If someone is ever just looking at the API for `MatlabAction.java` and doesn't pay enough attention to realize that `moveJsonArtifactToBuildRoot` is tied to one of two constructors it could be problematic.

Reply:
We have a `null` check in the getter [here](https://github.com/mathworks/jenkins-matlab-plugin/pull/354/files/becf595f879c406a4450007329bdbc79b2c2f302#diff-b207e06a879c72043c0244c0f0b181868e031a2b275c265cfc339ada14f7e881R24-R26). I saw Nikhil had done similar for `BuildArtifactAction.java` earlier.